scanpy/preprocessing/_normalization.py

```diff
@@ -11,14 +11,17 @@
 
 def _normalize_data(X, counts, after=None, copy=False):
     X = X.copy() if copy else X
-    after = np.median(counts[counts>0]) if after is None else after
+    if issubclass(X.dtype.type, (int, np.integer)):
+        X = X.astype(np.float32)  # TODO: Check if float64 should be used
```

Comment:
yeah, we should at least implement that TODO, leaking precision shouldn’t be done willy-nilly.

Unless that’s a huge performance cost of course.

Reply:
Oh, by check I meant do a benchmarking and seeing what range of values would be required for imprecision. Looking at the functions we're calling, it's all numpy and nothing is iterative on our end, so I'm not too worried about it. I think it'll be no worse than our default behavior.

What were you thinking for implementation? int64 -> float64 or something like that?

Reply:
Everything is simplified by the fact that 1) scRNASeq data is pretty robust to rounding [citation needed] and 2) neither counts nor any normalization we do go far up. it still feels weird:

- float32 can represent ints <= 16777216 precisely
- int64 can represent ints <= 9223372036854775807 precisely

I’d sleep more easily if I knew what order of magnitude the highest number has that exists at any point in scanpy (e.g. is there some point where we take count data to the power of something and save it)